### ISSUE TITLE
Lint/ShadowedException cop not working as expected for RestClient::RequestFailed exception

### ISSUE THREAD

**Issue by liijunwei** [rubocop/rubocop#10823]
Lint/ShadowedException cop not working as expected for RestClient::RequestFailed exception

--------

## Expected behavior

rubocop complains that `RestClient::MethodNotAllowed` was shadowed by `RestClient::RequestFailed`

## Actual behavior

rubocop did not complain

## Steps to reproduce the problem

```ruby
# be rubocop tmp.rb --only Lint/ShadowedException
require 'rest-client'
require 'json'

class Demo
  def run
    puts 'running...'

    hash = {}
    hash['RestClient::RequestFailed'] = RestClient::RequestFailed.ancestors
    hash['RestClient::MethodNotAllowed'] = RestClient::MethodNotAllowed.ancestors

    jj hash

    raise RestClient::MethodNotAllowed
  rescue RestClient::RequestFailed => e
    puts "#{__LINE__} #{e.class} #{e.message}"
  rescue RestClient::MethodNotAllowed => e
    puts "#{__LINE__} #{e.class} #{e.message}"
  end
end

Demo.new.run
```
This is extremely important! Providing us with a reliable way to reproduce
a problem will expedite its solution.

## RuboCop version

Include the output of `rubocop -V` or `bundle exec rubocop -V` if using Bundler.
If you see extension cop versions (e.g. `rubocop-performance`, `rubocop-rspec`, and others)
output by `rubocop -V`, include them as well. Here's an example:

```
$ [bundle exec] rubocop -V
1.31.2 (using Parser 2.7.2.0, rubocop-ast 1.1.1, running on ruby 2.7.2 x86_64-linux)
  - rubocop-performance 1.9.1
  - rubocop-rspec 2.0.0
```

**Comment by liijunwei** [comment]
pls ping me if any other necessary info is needed

**Comment by koic** [comment]
RuboCop is a static analysis tool, it doesn't know inheritance relationships at runtime. It has false negatives for unknown inheritance relationships. For example, the following checks cannot be performed at runtime:

```ruby
require 'rest-client'

RestClient::MethodNotAllowed.ancestors.include?(RestClient::RequestFailed) # => true
```

**Comment by liijunwei** [comment]
I get it now, thanks you 🙏 

https://en.wikipedia.org/wiki/Program_analysis
https://en.wikipedia.org/wiki/Static_program_analysis

**Comment by liijunwei** [comment]
Actually this could work

if I add `require 'rest-client'` after this line
https://github.com/rubocop/rubocop/blob/cbd4732028243dff01261fd3072b8f33f7771434/lib/rubocop/cop/lint/shadowed_exception.rb#L54

or add to this cop, I can check whether the exception is shadowed

**all I need to do is to make that `require` configurable**

**Comment by liijunwei** [comment]
solution for rails project:

```bash
bundle binstubs rubocop

# edit bin/rubocop

# add following lines to bin/rubocop(after  `require "bundler/setup"`)
require "rails/all" # might be required
Bundler.require

# then run
bin/rubocop --cache false --only Lint/ShadowedException
```

and then hopefully `runtime` exceptions can also be checked as `require "rails/all"` and `Bundler.require` import all the project dependencies, they should be available to be used in `Kernel.const_get`
